File: opennow-stable/src/main/gfn/persistentStorage.ts
import type {
  PersistentStorageLocation,
  PersistentStorageLocationsResult,
  PersistentStorageResetResult,
} from "@shared/gfn";
import {
  buildPaywallHeaders,
  GFN_PAYWALL_API_BASE_URL,
  parsePaywallMessage,
  readPaywallJson,
  resolvePaywallTokenCandidates,
} from "./paywall";

const GFN_STATUS_COMPONENTS_URL = "https://status.geforcenow.com/api/v2/components.json";
const DEFAULT_PAYWALL_LOCALE = "en_US";
const FALLBACK_FETCH_TIMEOUT_MS = 10_000;

interface ResetPersistentStorageInput {
  idToken: string;
  idTokenAlternates?: string[];
  storageRegion?: string | null;
}

interface FetchPersistentStorageLocationsInput {
  idToken: string;
  idTokenAlternates?: string[];
  vpcId?: string | null;
  locale?: string;
  currentRegionCode?: string | null;
  currentRegionName?: string | null;
}

interface PaywallProductsResponse {
  status?: unknown;
  products?: PaywallProduct[];
}

interface PaywallProduct {
  id?: unknown;
  productId?: unknown;
  productType?: unknown;
  add_on?: PaywallProduct[];
  regions?: PaywallStorageRegion[];
}

interface PaywallStorageRegion {
  metroRegion?: unknown;
  metroRegionName?: unknown;
  isAvailable?: unknown;
  isRecommendedRegion?: unknown;
}

interface StatuspageComponentsResponse {
  components?: StatuspageComponent[];
}

interface StatuspageComponent {
  id?: unknown;
  name?: unknown;
  status?: unknown;
  position?: unknown;
  group_id?: unknown;
  group?: unknown;
}

interface StorageStatusLocation {
  name: string;
  codes: string[];
  isAvailable?: boolean;
}

interface PaywallFailure {
  status: number;
  message: string;
  payload: unknown;
}

const FALLBACK_STORAGE_STATUS_LOCATIONS: StorageStatusLocation[] = [
  { name: "Northern California (USA)", codes: ["NP-SJC6-04", "NP-SJC6-06"] },
  { name: "Southern California (USA)", codes: ["NP-LAX-03"] },
  { name: "Oregon (USA)", codes: ["NP-PDX-01"] },
  { name: "Arizona (USA)", codes: ["NP-PHX-02"] },
  { name: "Texas (USA)", codes: ["NP-DAL-04", "NP-DAL-05", "NP-DAL-06"] },
  { name: "Illinois (USA)", codes: ["NP-CHI-04", "NP-CHI-05"] },
  { name: "Florida (USA)", codes: ["NP-MIA-03", "NP-MIA-04"] },
  { name: "Georgia (USA)", codes: ["NP-ATL-03", "NP-ATL-04"] },
  { name: "Virginia (USA)", codes: ["NP-ASH-04"] },
  { name: "New Jersey (USA)", codes: ["NP-NWK-03", "NP-NWK-04"] },
  { name: "Quebec (Canada)", codes: ["NP-MON-02"] },
  { name: "United Kingdom 1", codes: ["NP-LON-07", "NP-LON-08"] },
  { name: "Sweden", codes: ["NP-STH-03", "NP-STH-04"] },
  { name: "Netherlands North", codes: ["NP-AMS-07", "NP-AMS-06", "NP-AMS-08"] },
  { name: "Germany", codes: ["NP-FRK-06", "NP-FRK-07", "NP-FRK-08"] },
  { name: "France 1", codes: ["NP-PAR-05", "NP-PAR-06", "NP-PAR-07"] },
  { name: "Poland", codes: ["NP-WAW-01"] },
  { name: "Bulgaria", codes: ["NP-SOF-02"] },
  { name: "India", codes: ["NP-BOM-01"] },
  { name: "Japan", codes: ["NP-TYO-01"] },
];

function normalizeStorageRegion(storageRegion: string | null | undefined): string | null {
  if (typeof storageRegion !== "string") {
    return null;
  }

  const trimmed = storageRegion.trim();
  return trimmed.length > 0 ? trimmed : null;
}

function buildResetStorageUrl(storageRegion: string | null): string {
  const regionQueryValue = storageRegion ?? "null";
  return `${GFN_PAYWALL_API_BASE_URL}/reset/storage?storageRegion=${encodeURIComponent(regionQueryValue)}`;
}

function buildProductsUrl(input: Pick<FetchPersistentStorageLocationsInput, "locale" | "vpcId">): string {
  const url = new URL(`${GFN_PAYWALL_API_BASE_URL}/products`);
  url.searchParams.set("locale", input.locale?.trim() || DEFAULT_PAYWALL_LOCALE);
  if (input.vpcId?.trim()) {
    url.searchParams.set("vpcId", input.vpcId.trim());
  }
  return url.toString();
}

function asString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim().length > 0 ? value.trim() : undefined;
}

function normalizeRegionCode(value: string | null | undefined): string | undefined {
  return typeof value === "string" && value.trim().length > 0 ? value.trim() : undefined;
}

function normalizeRegionName(value: string | null | undefined): string | undefined {
  return typeof value === "string" && value.trim().length > 0 ? value.trim() : undefined;
}

function normalizeStatusComponentName(value: string): string {
  return value
    .replace(/\s*\[[^\]]+\]\s*/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function normalizeStatusCode(value: string): string | undefined {
  const normalized = normalizeStatusComponentName(value);
  return /^NPA?-[A-Z0-9]+(?:-[A-Z0-9]+)*$/.test(normalized) ? normalized : undefined;
}

function isCurrentLocationName(name: string, currentRegionName: string | undefined): boolean {
  return Boolean(currentRegionName && name.toLowerCase() === currentRegionName.toLowerCase());
}

function isAuthFailure(failure: PaywallFailure): boolean {
  return (
    failure.status === 401 ||
    failure.status === 403 ||
    /idtoken|unauthori[sz]ed|forbidden|token/i.test(failure.message)
  );
}

function isRegionFailure(failure: PaywallFailure): boolean {
  return /region|storageRegion|products not found/i.test(failure.message);
}

function paywallFailureMessage(failure: PaywallFailure | null, fallback: string): string {
  if (!failure) {
    return fallback;
  }
  if (/starfleet idtoken was invalid/i.test(failure.message)) {
    return "NVIDIA rejected the saved session token for this storage API. Please log in again, then retry.";
  }
  return failure.message;
}

function isStorageProductCandidate(product: PaywallProduct): boolean {
  const productType = asString(product.productType)?.toUpperCase();
  return productType === "STORAGE" || productType === "PAID" || Array.isArray(product.regions);
}

function collectStorageProductCandidates(products: PaywallProduct[]): PaywallProduct[] {
  const candidates: PaywallProduct[] = [];

  for (const product of products) {
    if (isStorageProductCandidate(product)) {
      candidates.push(product);
    }

    if (Array.isArray(product.add_on)) {
      for (const addon of product.add_on) {
        if (isStorageProductCandidate(addon)) {
          candidates.push(addon);
        }
      }
    }
  }

  return candidates;
}

function chooseStorageRegions(products: PaywallProduct[]): PaywallStorageRegion[] {
  const candidates = collectStorageProductCandidates(products);
  const selected = candidates.find((product) => Array.isArray(product.regions) && product.regions.length > 0);
  return selected?.regions ?? [];
}

function toLocation(
  region: PaywallStorageRegion,
  currentRegionCode: string | undefined,
): PersistentStorageLocation | null {
  const code = asString(region.metroRegion);
  if (!code) {
    return null;
  }

  const name = asString(region.metroRegionName) ?? code;
  return {
    code,
    name,
    isAvailable: region.isAvailable !== false,
    isCurrent: currentRegionCode ? code === currentRegionCode : undefined,
    isRecommended: region.isRecommendedRegion === true,
  };
}

function dedupeLocations(locations: PersistentStorageLocation[]): PersistentStorageLocation[] {
  const seen = new Set<string>();
  const deduped: PersistentStorageLocation[] = [];

  for (const location of locations) {
    if (seen.has(location.code)) {
      continue;
    }
    seen.add(location.code);
    deduped.push(location);
  }

  return deduped;
}

function dedupeCodes(codes: string[]): string[] {
  const seen = new Set<string>();
  const deduped: string[] = [];

  for (const code of codes) {
    const normalized = normalizeRegionCode(code);
    if (!normalized || seen.has(normalized)) {
      continue;
    }
    seen.add(normalized);
    deduped.push(normalized);
  }

  return deduped;
}

function ensureCurrentLocation(
  locations: PersistentStorageLocation[],
  currentRegionCode: string | undefined,
  currentRegionName: string | undefined,
): PersistentStorageLocation[] {
  if (!currentRegionCode || locations.some((location) => location.code === currentRegionCode)) {
    return locations;
  }

  return [
    {
      code: currentRegionCode,
      name: currentRegionName ?? currentRegionCode,
      isAvailable: true,
      isCurrent: true,
    },
    ...locations,
  ];
}

function statusLocationsToPersistentLocations(
  statusLocations: StorageStatusLocation[],
  currentRegionCode: string | undefined,
  currentRegionName: string | undefined,
): PersistentStorageLocation[] {
  const locations = statusLocations
    .map<PersistentStorageLocation | null>((location) => {
      const codes = dedupeCodes(location.codes);
      if (codes.length === 0) {
        return null;
      }

      const currentByName = isCurrentLocationName(location.name, currentRegionName);
      const currentCode = currentByName && currentRegionCode ? currentRegionCode : undefined;
      const code = currentCode ?? codes[0];

      return {
        code,
        name: location.name,
        isAvailable: location.isAvailable !== false,
        isCurrent: currentRegionCode ? code === currentRegionCode || currentByName : currentByName || undefined,
      };
    })
    .filter((location): location is PersistentStorageLocation => Boolean(location));

  return ensureCurrentLocation(dedupeLocations(locations), currentRegionCode, currentRegionName);
}

function fallbackPersistentStorageLocations(
  currentRegionCode: string | undefined,
  currentRegionName: string | undefined,
): PersistentStorageLocation[] {
  return statusLocationsToPersistentLocations(
    FALLBACK_STORAGE_STATUS_LOCATIONS,
    currentRegionCode,
    currentRegionName,
  );
}

async function fetchWithTimeout(url: string, options: RequestInit = {}): Promise<Response> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), FALLBACK_FETCH_TIMEOUT_MS);
  try {
    return await fetch(url, {
      ...options,
      signal: controller.signal,
    });
  } finally {
    clearTimeout(timeout);
  }
}

function parseStatuspageStorageLocations(payload: unknown): StorageStatusLocation[] {
  const response = payload as StatuspageComponentsResponse;
  const components = Array.isArray(response.components) ? response.components : [];
  const groups = new Map<string, StatuspageComponent>();

  for (const component of components) {
    const id = asString(component.id);
    if (id && component.group === true) {
      groups.set(id, component);
    }
  }

  const storageLocations: StorageStatusLocation[] = [];
  const sortedGroups = [...groups.values()].sort((a, b) => {
    const left = typeof a.position === "number" ? a.position : Number.MAX_SAFE_INTEGER;
    const right = typeof b.position === "number" ? b.position : Number.MAX_SAFE_INTEGER;
    return left - right;
  });

  for (const group of sortedGroups) {
    const groupId = asString(group.id);
    const groupName = asString(group.name);
    if (!groupId || !groupName) {
      continue;
    }

    const children = components
      .filter((component) => asString(component.group_id) === groupId)
      .sort((a, b) => {
        const left = typeof a.position === "number" ? a.position : Number.MAX_SAFE_INTEGER;
        const right = typeof b.position === "number" ? b.position : Number.MAX_SAFE_INTEGER;
        return left - right;
      });
    const storageChild = children.find((child) => asString(child.name) === "Cloud Storage");
    if (!storageChild) {
      continue;
    }

    const codes = dedupeCodes(
      children
        .map((child) => asString(child.name))
        .map((name) => name ? normalizeStatusCode(name) : undefined)
        .filter((code): code is string => Boolean(code)),
    );
    if (codes.length === 0) {
      continue;
    }

    storageLocations.push({
      name: normalizeStatusComponentName(groupName),
      codes,
      isAvailable: storageChild.status !== "major_outage",
    });
  }

  return storageLocations;
}

async function fetchStatuspageStorageLocations(
  currentRegionCode: string | undefined,
  currentRegionName: string | undefined,
): Promise<PersistentStorageLocation[]> {
  try {
    const response = await fetchWithTimeout(GFN_STATUS_COMPONENTS_URL, {
      headers: { Accept: "application/json" },
    });
    if (!response.ok) {
      return fallbackPersistentStorageLocations(currentRegionCode, currentRegionName);
    }

    const payload = await readPaywallJson(response);
    const statusLocations = parseStatuspageStorageLocations(payload);
    if (statusLocations.length === 0) {
      return fallbackPersistentStorageLocations(currentRegionCode, currentRegionName);
    }

    return statusLocationsToPersistentLocations(statusLocations, currentRegionCode, currentRegionName);
  } catch {
    return fallbackPersistentStorageLocations(currentRegionCode, currentRegionName);
  }
}

function getFallbackRegionCandidates(storageRegion: string | null): Array<string | null> {
  if (!storageRegion) {
    return [null];
  }

  const normalizedRegion = normalizeRegionCode(storageRegion);
  if (!normalizedRegion) {
    return [storageRegion];
  }

  const location = FALLBACK_STORAGE_STATUS_LOCATIONS.find((candidate) =>
    candidate.codes.some((code) => code === normalizedRegion),
  );
  if (!location) {
    return [normalizedRegion];
  }

  return dedupeCodes([normalizedRegion, ...location.codes]);
}

export async function fetchPersistentStorageLocations(
  input: FetchPersistentStorageLocationsInput,
): Promise<PersistentStorageLocationsResult> {
  const currentRegionCode = normalizeRegionCode(input.currentRegionCode);
  const currentRegionName = normalizeRegionName(input.currentRegionName);
  let lastFailure: PaywallFailure | null = null;

  for (const idToken of await resolvePaywallTokenCandidates(input.idToken, input.idTokenAlternates)) {
    const response = await fetch(buildProductsUrl(input), {
      headers: buildPaywallHeaders(idToken),
    });

    const payload = await readPaywallJson(response);
    if (!response.ok) {
      lastFailure = {
        status: response.status,
        message: parsePaywallMessage(payload) ?? `Persistent storage locations failed with status ${response.status}`,
        payload,
      };
      if (isAuthFailure(lastFailure)) {
        continue;
      }
      break;
    }

    const productsResponse = payload as PaywallProductsResponse;
    if (productsResponse.status === "failure") {
      lastFailure = {
        status: response.status,
        message: parsePaywallMessage(productsResponse) ?? "Persistent storage locations request failed",
        payload,
      };
      if (isAuthFailure(lastFailure)) {
        continue;
      }
      break;
    }

    const products = Array.isArray(productsResponse.products) ? productsResponse.products : [];
    const locations = ensureCurrentLocation(
      dedupeLocations(
        chooseStorageRegions(products)
          .map((region) => toLocation(region, currentRegionCode))
          .filter((location): location is PersistentStorageLocation => Boolean(location)),
      ),
      currentRegionCode,
      currentRegionName,
    );

    if (locations.length > 0) {
      return {
        locations,
        currentRegionCode,
        currentRegionName,
      };
    }
  }

  if (lastFailure) {
    console.warn("Persistent storage locations fell back to status page:", lastFailure.message);
  }

  return {
    locations: await fetchStatuspageStorageLocations(currentRegionCode, currentRegionName),
    currentRegionCode,
    currentRegionName,
  };
}

export async function resetPersistentStorage(
  input: ResetPersistentStorageInput,
): Promise<PersistentStorageResetResult> {
  const storageRegion = normalizeStorageRegion(input.storageRegion);
  const storageRegions = getFallbackRegionCandidates(storageRegion);
  let lastFailure: PaywallFailure | null = null;

  for (const idToken of await resolvePaywallTokenCandidates(input.idToken, input.idTokenAlternates)) {
    for (const candidateRegion of storageRegions) {
      const response = await fetch(buildResetStorageUrl(candidateRegion), {
        method: "POST",
        headers: buildPaywallHeaders(idToken),
        body: null,
      });

      const payload = await readPaywallJson(response);
      if (response.ok) {
        return {
          ok: true,
          storageRegion: candidateRegion,
          message: parsePaywallMessage(payload),
        };
      }

      lastFailure = {
        status: response.status,
        message: parsePaywallMessage(payload) ?? `Persistent storage reset failed with status ${response.status}`,
        payload,
      };

      if (isAuthFailure(lastFailure)) {
        break;
      }
      if (!isRegionFailure(lastFailure)) {
        throw new Error(paywallFailureMessage(lastFailure, "Persistent storage reset failed"));
      }
    }
  }

  throw new Error(paywallFailureMessage(lastFailure, "Persistent storage reset failed"));
}

File: native/opennow-core/src/persistent_storage.rs
use crate::gfn::{AuthSession, ServiceError};
use reqwest::blocking::Client;
use reqwest::header::{ACCEPT, CONTENT_TYPE, HeaderMap, HeaderValue};
use serde_json::{Value, json};
use url::Url;

const PAYWALL_BASE: &str = "https://api-prod.nvidia.com/gfn-paywall-api/api/v2";

pub struct PersistentStorageService {
    client: Client,
}

impl PersistentStorageService {
    pub fn new(client: Client) -> Self {
        Self { client }
    }

    pub fn locations(&self, params: &Value, auth: &AuthSession) -> Result<Value, ServiceError> {
        let current_code = params["currentRegionCode"]
            .as_str()
            .filter(|value| !value.trim().is_empty());
        let current_name = params["currentRegionName"]
            .as_str()
            .filter(|value| !value.trim().is_empty());
        let mut url =
            Url::parse(&format!("{PAYWALL_BASE}/products")).expect("constant paywall URL is valid");
        url.query_pairs_mut()
            .append_pair("locale", params["locale"].as_str().unwrap_or("en_US"));
        if let Some(vpc_id) = params["serverRegionId"]
            .as_str()
            .filter(|value| !value.trim().is_empty())
        {
            url.query_pairs_mut().append_pair("vpcId", vpc_id);
        }
        let token = session_token(auth)?;
        let response = self.client.get(url).headers(paywall_headers(token)?).send();
        let mut locations = response
            .ok()
            .and_then(|response| {
                if !response.status().is_success() {
                    return None;
                }
                response
                    .json::<Value>()
                    .ok()
                    .map(|payload| locations_from_products(&payload, current_code))
            })
            .unwrap_or_default();
        if locations.is_empty() {
            locations = fallback_locations(current_code, current_name);
        }
        if let Some(code) = current_code
            && !locations
                .iter()
                .any(|location| location["code"].as_str() == Some(code))
        {
            locations.insert(0, json!({"code":code,"name":current_name.unwrap_or(code),"isAvailable":true,"isCurrent":true}));
        }
        Ok(
            json!({"locations":locations,"currentRegionCode":current_code,"currentRegionName":current_name}),
        )
    }

    pub fn reset(&self, params: &Value, auth: &AuthSession) -> Result<Value, ServiceError> {
        if params["confirmed"].as_bool() != Some(true) {
            return Err(ServiceError {
                code: "confirmation_required",
                message: "Persistent storage reset requires explicit confirmation".to_owned(),
            });
        }
        let region = params["storageRegion"]
            .as_str()
            .filter(|value| !value.trim().is_empty());
        let mut url = Url::parse(&format!("{PAYWALL_BASE}/reset/storage"))
            .expect("constant paywall URL is valid");
        url.query_pairs_mut()
            .append_pair("storageRegion", region.unwrap_or("null"));
        let response = self
            .client
            .post(url)
            .headers(paywall_headers(session_token(auth)?)?)
            .send()
            .map_err(|error| ServiceError {
                code: "network_error",
                message: format!("Persistent storage reset failed: {error}"),
            })?;
        let status = response.status();
        let payload = response.json::<Value>().unwrap_or(Value::Null);
        if !status.is_success() {
            let message = payload["message"]
                .as_str()
                .or_else(|| payload["errors"]["errorMessage"].as_str())
                .unwrap_or("Persistent storage reset failed");
            let message = if message.contains("idToken") {
                "NVIDIA requires a web-account session for this storage action. Open NVIDIA Storage Manager in your browser."
            } else {
                message
            };
            return Err(ServiceError {
                code: if matches!(status.as_u16(), 401 | 403) {
                    "authentication_required"
                } else {
                    "upstream_error"
                },
                message: message.to_owned(),
            });
        }
        Ok(json!({"ok":true,"storageRegion":region,"message":payload["message"]}))
    }
}

fn locations_from_products(payload: &Value, current_code: Option<&str>) -> Vec<Value> {
    let mut products = payload["products"].as_array().cloned().unwrap_or_default();
    let addons = products
        .iter()
        .flat_map(|product| product["add_on"].as_array().cloned().unwrap_or_default())
        .collect::<Vec<_>>();
    products.extend(addons);
    let regions = products
        .iter()
        .find_map(|product| {
            let kind = product["productType"].as_str().unwrap_or_default();
            (kind.eq_ignore_ascii_case("STORAGE")
                || kind.eq_ignore_ascii_case("PAID")
                || product["regions"].is_array())
            .then(|| product["regions"].as_array().cloned())
            .flatten()
            .filter(|values| !values.is_empty())
        })
        .unwrap_or_default();
    let mut seen = std::collections::HashSet::new();
    regions.into_iter().filter_map(|region| {
        let code = region["metroRegion"].as_str()?.trim();
        if code.is_empty() || !seen.insert(code.to_owned()) { return None; }
        Some(json!({
            "code":code,"name":region["metroRegionName"].as_str().unwrap_or(code),
            "isAvailable":region["isAvailable"].as_bool().unwrap_or(true),
            "isCurrent":current_code == Some(code),"isRecommended":region["isRecommendedRegion"].as_bool().unwrap_or(false)
        }))
    }).collect()
}

fn fallback_locations(current_code: Option<&str>, current_name: Option<&str>) -> Vec<Value> {
    const LOCATIONS: [(&str, &str); 20] = [
        ("NP-SJC6-04", "Northern California (USA)"),
        ("NP-LAX-03", "Southern California (USA)"),
        ("NP-PDX-01", "Oregon (USA)"),
        ("NP-PHX-02", "Arizona (USA)"),
        ("NP-DAL-04", "Texas (USA)"),
        ("NP-CHI-04", "Illinois (USA)"),
        ("NP-MIA-03", "Florida (USA)"),
        ("NP-ATL-03", "Georgia (USA)"),
        ("NP-ASH-04", "Virginia (USA)"),
        ("NP-NWK-03", "New Jersey (USA)"),
        ("NP-MON-02", "Quebec (Canada)"),
        ("NP-LON-07", "United Kingdom"),
        ("NP-STH-03", "Sweden"),
        ("NP-AMS-07", "Netherlands North"),
        ("NP-FRK-06", "Germany"),
        ("NP-PAR-05", "France"),
        ("NP-WAW-01", "Poland"),
        ("NP-SOF-02", "Bulgaria"),
        ("NP-BOM-01", "India"),
        ("NP-TYO-01", "Japan"),
    ];
    LOCATIONS.into_iter().map(|(code,name)| json!({
        "code":code,"name":if current_code == Some(code) { current_name.unwrap_or(name) } else { name },
        "isAvailable":true,"isCurrent":current_code == Some(code),"isRecommended":false
    })).collect()
}

fn paywall_headers(token: &str) -> Result<HeaderMap, ServiceError> {
    let mut headers = HeaderMap::new();
    headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
    headers.insert(
        "idtoken",
        HeaderValue::from_str(token).map_err(|_| ServiceError {
            code: "invalid_params",
            message: "Invalid authentication token".to_owned(),
        })?,
    );
    Ok(headers)
}

fn session_token(auth: &AuthSession) -> Result<&str, ServiceError> {
    auth.tokens
        .id_token
        .as_deref()
        .or(Some(auth.tokens.access_token.as_str()))
        .filter(|value| !value.is_empty())
        .ok_or_else(|| ServiceError {
            code: "authentication_required",
            message: "No authenticated token is available".to_owned(),
        })
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn product_regions_are_deduplicated() {
        let payload = json!({"products":[{"productType":"STORAGE","regions":[
            {"metroRegion":"NP-AMS-07","metroRegionName":"Amsterdam","isAvailable":true},
            {"metroRegion":"NP-AMS-07","metroRegionName":"Duplicate"}
        ]}]});
        let result = locations_from_products(&payload, Some("NP-AMS-07"));
        assert_eq!(result.len(), 1);
        assert_eq!(result[0]["isCurrent"], true);
    }
}
